### ISSUE TITLE
GitHub 25 | MagiAttention: A Distributed Attention Towards Linear Scalability for Ultra-Long Context, Heterogeneous Mask Training

### ISSUE THREAD

**Issue by NightDreamy** [AI-Infra-Team/awesome-papers#58]
## MagiAttention：A Distributed Attention Towards Linear Scalability for Ultra-Long Context, Heterogeneous Mask Training

Zewei, Tao and Yunpeng, Huang

SandAI, Nanjing University

https://github.com/SandAI-org/MagiAttention

![Image](https://github.com/user-attachments/assets/e01c4fc3-9314-413b-a435-9980c95794cc)

### Motivation

Training large-scale models for video generation presents two major challenges: 

(1) The extremely long context length of video tokens, which reaching up to 4 million during training, results in prohibitive computational and memory overhead. Context-Parallelism (CP) is designed for dealing such long context challenge, but existing state-of-the-art CP methods  face scalability limitations that face scalability limitations due to size constraints or the high communication overhead inherent in inefficient ring-style point-to-point (P2P) patterns. 

(2) The combination of block-causal attention and Packing-and-Padding (PnP) introduces highly complex attention mask patterns with variable sequence lengths, which cannot be efficiently handled by existing attention implementations. Second, the communication overhead issue is exacerbated under sparse varlen mask settings, as entire sequence chunks are still transferred across all CP ranks even when not all ranks require them, might causing over 30% redundant communication costs as illustrated in Zero-Redundant Comm. Third, the former challenges cause the pipeline compute-communication overlap strategy fails more often due to imbalanced loads and large communication overheads, further limiting scalability.

![img](https://sandai-org.github.io/MagiAttention/assets/img/magiattn/comp/ring_attn_load_balance.png)

​	While recent efforts dynamically adjust CP sizes to avoid unnecessary sharding and redundant communication for shorter sequences, they still incur extra memory overhead for NCCL buffers and involve complex scheduling to balance loads and synchronize across different subsets of ranks.

![Image](https://github.com/user-attachments/assets/683a688e-d411-4817-bbf5-0d3dfad2e895)

​	To address these challenges, this work proposes MagiAttention, which aims to support a wide variety of attention mask types with **kernel-level flexibility**, while achieving **linear scalability** with respect to context-parallel (CP) size across a broad range of scenarios, particularly suitable for training tasks involving *ultra-long, heterogeneous mask* training like video-generation for Magi-1.

#### Flex-Flash-Attn

we introduce Flex-Flash-Attention (FFA), which is natively designed for distribution scenarios and provides greater flexibility in handling diverse attention mask types. The core idea behind FFA is to generalize a **distributable** formulation for irregular attention masks by decomposing the entire mask into multiple computational units, each referred to as an AttnSlice. Each AttnSlice is defined by a triplet (QRange,KRange,MaskType), which specifies a submask with a basic shape bounded by a contiguous 2D query-key region as seen in the figure below.

![Image](https://github.com/user-attachments/assets/e597eaff-3c34-428c-be19-5045817691f7)

Although AttnSlice can represent most mask patterns using two common types (FULL and CAUSAL), it is inefficient for patterns like sliding-window, which requires row-by-row expression. To address this, we introduce two new mask types, INV-CAUSAL and BI-CAUSAL, to efficiently represent more specific sliding-window-style patterns. We provide basic examples of these four mask types in the following figures.

![Image](https://github.com/user-attachments/assets/2177ea93-8b89-4f03-a225-23d179522c61)

![Image](https://github.com/user-attachments/assets/ff4c6ad2-92b4-46bb-8534-664eaf7fefbb)

#### Comp Load-Balance Comp

First, to enable finer-grained control, we propose a chunk-wise permutable sharding strategy as seen in Overview. Specifically, the entire mask is evenly partitioned along the query-dimension into chunks, each associated with a submask area: {(Ci, Area(Ci))}i=1-n, where Ci indicates i-th chunk, Area(Ci) is the mask area of Ci, n is seqlenchunk_size, and chunk_size is a hyperparameter controlling granularity.

These chunks are then equally assigned to cp_size buckets, with each bucket containing the exact same number of chunks to ensure token-level load balance in non-attention modules, attaching with a summed submask area, denoted as {(Bj, SumArea(Bj))}j=1-cp_size.

With above strategy, we could fine-grained control the computational workloads of each CP rank, and the load-balancing dispatch becomes a combinatorial optimization problem, defined as finding an optimal mapping function f∗: {Ci}n=1-n → {Bj}j=1-cp_size

However, this optimization is a known NP-hard problem, making it impractical to find an optimal solution on-the-fly during each training iteration, especially given the varying mask patterns across micro-batches. Thus, we propose an efficient greedy algorithm as shown below that provides a suboptimal yet effective solution within O(nlog⁡n) complexity.

#### Zero-Redundant Comm

To address this, as illustrated in the figure below, we introduce two communication primitives: **Group-Cast** and **Group-Reduce**, which model the communication patterns of low-demand KV and dKV. For example, in the causal mask, KV5 on rank2 is required only by Q6,Q7 and should be sent exclusively to the target ranks rank0,rank1 via Group-Cast, while the partial dKV5 is collected and reduced back to rank2 via Group-Reduce accordingly.

![Image](https://github.com/user-attachments/assets/582f2d8a-23c6-4daa-967d-f229c9468436)

#### Multi-Stage Overlap

Leveraging previous optimizations, we achieve high-performance computation through an efficient kernel and balanced workload dispatch, while minimizing communication overhead with our new primitives. To drive true linear scalability, we further improve end-to-end performance by introducing a multi-stage compute-communication overlap strategy, that effectively hides communication latency and adaptively optimizes overlap through manual or automatic tuning.

we schedule pipeline stages to overlap computation with communication for both forward and backward passes, as shown in the following figureFig. Each rank*i* first partitions its remote KV/dKV communication into stages. In the pass, the scheduler first launches the group-cast kernel to prefetch the next remote KV, then asynchronously executes the FFA kernel for partial attention computation, hiding all communication behind computation 7 . In the backward pass, besides prefetching the next KV, the group-reduce kernel reduces the last dKV in a separate CUDA stream before launching the FFA kernel for the current stage, ensuring communication is overlapped across all stages except the final dKV reduction.

![Image](https://github.com/user-attachments/assets/c675bda4-0540-4904-8756-2c3cfe7657df)
